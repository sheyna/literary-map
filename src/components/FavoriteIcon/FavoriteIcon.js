import { useState, useEffect } from 'react';
import { useAuth0 } from "@auth0/auth0-react";
import axios from 'axios';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faHeart } from '@fortawesome/free-solid-svg-icons';
import PropTypes from 'prop-types';

import styles from './FavoriteIcon.module.css';

function FavoriteIcon({locationInfo, favorites, setFavorites}) {
  const { isAuthenticated, getIdTokenClaims } = useAuth0();
	const [ showFavoriteToolTip, setShowFavoriteToolTip ] = useState(false);
	const [ favIconHover, setFavIconHover ] = useState(false);


	const [ isFavoritedLocation, setIsFavoritedLocation ] = useState(false);
	const [ favLocationObj, setFavLocationObj ] = useState({})

	useEffect(function() {
		const favoriteLocationObj = favorites.length && favorites.find(loc => loc.locationId === locationInfo.locationId);
		const favorited = favoriteLocationObj ? true : false;
		setFavLocationObj(favoriteLocationObj);
		setIsFavoritedLocation(favorited);
	}, [favorites, locationInfo.locationId]);

  async function addFavorites(locationToAdd) {
    if (isAuthenticated) {
      try {
        // get token:
        const res = await getIdTokenClaims();
        const jwt = res.__raw;
  
        const config = {
          method: 'post',
          baseURL: process.env.REACT_APP_SERVER,
          url: '/favorites',
          headers: {"Authorization": `Bearer ${jwt}`},
          data: locationToAdd
        }
        const newFav = await axios(config);
        setFavLocationObj(newFav.data);
        setFavorites((fav) => [...fav, newFav.data]);
      } catch(err) {
        console.log('we have an error: ', err.response.data)
      }
    } 
  }
  
  async function deleteFavorites(id) {
    if (isAuthenticated) {
      try {
        // get token:
        const res = await getIdTokenClaims();
        const jwt = res.__raw;
  
        const config = {
          method: 'delete',
          baseURL: process.env.REACT_APP_SERVER,
          url: `/favorites/${id}`,
          headers: {"Authorization": `Bearer ${jwt}`}
        }
        await axios(config);
        let updatedFavs = favorites.filter(loc => loc._id !== id);
        setFavorites(updatedFavs);
      } catch(err) {
        console.log('we have an error: ', err.response.data)
      }
    } 
  }

	async function handleFavoriteChange() {
		if (isAuthenticated) {
			if (isFavoritedLocation) {
				setIsFavoritedLocation(false);
        deleteFavorites(favLocationObj._id);
				// favoriteModule.deleteFavorites(favLocationObj._id, setFavorites, isAuthenticated, getIdTokenClaims, favorites);
			} else {
				setIsFavoritedLocation(true);
        addFavorites(locationInfo);
				// favoriteModule.addFavorites(infoBoxLocation, setFavorites, isAuthenticated, getIdTokenClaims);
			} 
		} else {
			setShowFavoriteToolTip((show) => !show);
		}
	}

	// Re:hover inline CSS:
	// https://stackoverflow.com/questions/28365233/inline-css-styles-in-react-how-to-implement-ahover

	function toggleHover() {
		setFavIconHover((tog) => !tog);
	}

	const favoriteIconStyles = {
		color: favIconHover ? '#d88bb8' : isFavoritedLocation ? '#bd4089' : 'gray',
		transition: 'color 1s ease',
	};
  
  return (
    <div
      onClick={handleFavoriteChange}
    >
      <FontAwesomeIcon
        className={styles.menuIcon}
        style={favoriteIconStyles}
        icon={faHeart}
        onMouseEnter={toggleHover}
        onMouseLeave={toggleHover}
      />
      {
        showFavoriteToolTip
        && <div className={styles.tooltip}>
          <div className={styles.box}>
            <p>Please login to favorite</p>
          </div>
        </div>
      }
    </div>
  )
}

FavoriteIcon.propTypes = {
	locationInfo: PropTypes.shape({
		title: PropTypes.string,
		book: PropTypes.string,
		author: PropTypes.string,
		thumb: PropTypes.string,
		image: PropTypes.string,
		positionTop: PropTypes.number,
		positionLeft: PropTypes.number,
		townName: PropTypes.string,
		bodyText: PropTypes.string,
	}),
	favorites: PropTypes.array,
	setFavorites: PropTypes.func
};

export default FavoriteIcon
