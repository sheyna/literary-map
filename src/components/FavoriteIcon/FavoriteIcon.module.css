.tooltip {
	position: absolute;
	top: 3rem;
	right: 0;
}

.tooltip .box {
	background-color: #bd4089;
	position: relative;
	padding: 1rem;
	z-index: 300;
	color: white;
	font-weight: 600;
}

.tooltip::before {
	content: '';
	width: 0;
	height: 0;
	border-style: solid;
	border-width: 0 15px 50px 15px;
	border-color: transparent transparent #bd4089  transparent;
	position: absolute;
	bottom: 4.5rem;
	right: 0;
}

.menuIcon {
	font-size: 1.5rem;
	cursor: pointer;
}
