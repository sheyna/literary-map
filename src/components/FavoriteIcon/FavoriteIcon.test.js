import ReactTestRenderer from 'react-test-renderer';
import FavoriteIcon from  './FavoriteIcon';

const mockFn = jest.fn();

const testLocation = {
	"locationId": 9,
	"title": "Elizabeth Gaskell's Home",
	"book": "Mary Barton, North and South, Cranford, Wives and Daughters",
	"author": "Elizabeth Gaskell",
	"positionTop": 38,
	"positionLeft": 48,
	"townName":  "Manchester",
	"bodyText": "Gaskell's historic home was restored in 2014. It is now a museum to the author and her times."
};

const testLocationInFavs = [{...testLocation, _id: 5, __v: 0}];
const testLocationNotInFavs = [];

// {locationInfo, favorites, setFavorites}
describe('FavoriteIcon component', () => {
  it('should match snapshot of favs', () => {
    const component = ReactTestRenderer.create(
      <FavoriteIcon
        locationInfo={testLocation}
        favorites={testLocationInFavs}
        setFavorites={mockFn}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('should match snapshot of NOT favs', () => {
    const component = ReactTestRenderer.create(
      <FavoriteIcon
        locationInfo={testLocation}
        favorites={testLocationNotInFavs}
        setFavorites={mockFn}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
