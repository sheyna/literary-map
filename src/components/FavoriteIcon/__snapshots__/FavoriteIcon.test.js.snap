// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteIcon component should match snapshot of NOT favs 1`] = `
<div
  onClick={[Function]}
>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-heart menuIcon"
    data-icon="heart"
    data-prefix="fas"
    focusable="false"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    role="img"
    style={
      Object {
        "color": "gray",
        "transition": "color 1s ease",
      }
    }
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M47.6 300.4L228.3 469.1c7.5 7 17.4 10.9 27.7 10.9s20.2-3.9 27.7-10.9L464.4 300.4c30.4-28.3 47.6-68 47.6-109.5v-5.8c0-69.9-50.5-129.5-119.4-141C347 36.5 300.6 51.4 268 84L256 96 244 84c-32.6-32.6-79-47.5-124.6-39.9C50.5 55.6 0 115.2 0 185.1v5.8c0 41.5 17.2 81.2 47.6 109.5z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
</div>
`;

exports[`FavoriteIcon component should match snapshot of favs 1`] = `
<div
  onClick={[Function]}
>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-heart menuIcon"
    data-icon="heart"
    data-prefix="fas"
    focusable="false"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    role="img"
    style={
      Object {
        "color": "gray",
        "transition": "color 1s ease",
      }
    }
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M47.6 300.4L228.3 469.1c7.5 7 17.4 10.9 27.7 10.9s20.2-3.9 27.7-10.9L464.4 300.4c30.4-28.3 47.6-68 47.6-109.5v-5.8c0-69.9-50.5-129.5-119.4-141C347 36.5 300.6 51.4 268 84L256 96 244 84c-32.6-32.6-79-47.5-124.6-39.9C50.5 55.6 0 115.2 0 185.1v5.8c0 41.5 17.2 81.2 47.6 109.5z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
</div>
`;
